add non-graceful node shutdown KEP (#1116)
kubernetes/enhancements: keps/sig-storage/2268-non-graceful-shutdown/README.md

---
title: Non graceful shutdown
authors:
  - "@yastij"
owning-sig: sig-storage
participating-sigs:
  - sig-node
approvers:
  - "@saad-ali"
  - "@thockin"
reviewers:
  - "@andrewsykim"
  - "@xing-yang"
editor: Yassine Tijani
creation-date: 2019-06-26
last-updated: 2020-01-16
status: implementable
---

# Non graceful shutdown

This includes the Summary and Motivation sections.

## Table of Contents

<!-- toc -->
- [Release Signoff Checklist](#release-signoff-checklist)
- [Summary](#summary)
- [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals](#non-goals)
- [Proposal](#proposal)
  - [Risks and Mitigations](#risks-and-mitigations)
- [Design Details](#design-details)
  - [Test Plan](#test-plan)
  - [Graduation Criteria](#graduation-criteria)
  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
- [Implementation History](#implementation-history)
- [Alternatives](#alternatives)
  - [rely on the taint as a lock mechanism (taint and untaint by the same actor)](#rely-on-the-taint-as-a-lock-mechanism-taint-and-untaint-by-the-same-actor)
<!-- /toc -->

> 

Reviewer: Need a section for Production Readiness Review.  The `feature enablement and rollback` section is required for alpha:

https://github.com/kubernetes/enhancements/tree/master/keps/NNNN-kep-template#feature-enablement-and-rollback

* enable detach for all the storage providers

## Proposal

user stories:

* as a cluster administrator I want my stateful workload to failover in case of a node is partitioned without any intervention when it's safe to do so

* as a developer I would like to rely on a self-healing platform for my stateful workloads

This KEP plans to introduce the following capability to the CSI capability to the spec:

```go
// CSIDriverSpec is the specification of a CSIDriver.
type CSIDriverSpec struct {
  ...

	// SafeDetach indicates this CSI volume driver is able to perform detach
  // operations, when set to true, the CSI volume driver needs to ensure that 
  // ControllerUnpublishVolume() implementation is checking for detach safety
	// +optional
  SafeDetach *bool
}

```

the kube-controller-manager, would need to ensure that eviction looks at the `CSIDriver` and when `.spec.SafeDetach` is set true, forcefully delete pods.

Specifically the pod gc controller (the only controller able to forcefully delete pods) would also check now for stateful pods, and select those that satisfy ALL of the following conditions:

- pods that are backed by a provisioner that support safe detach

> - pods that are backed by volumes that aren't marked as `ReadWriteMany, this would avoid breaking having multiple replicas of the same pod.

Reviewer: In the existing pod gc controller, there are 3 routines that deletes pods:
- gcTerminated
- gcOrphaned
- gcUnscheduledTerminating

Are you going to modify the existing gcTerminated() to check this `SafeDetach` flag or does this need a new gcXXXX routine?  The KEP didn't describe what triggers the pod gc controller to delete pods other than checking the `SafeDetch` flag and `ReadWriteMany` volume.  Does it go thru the pods and check if pod.Status.Phase is not one of v1.PodPending, v1.PodRunning, and v1.PodUnknown and decide it should be terminated or does it need to check node status?  Please clarify.

Reviewer 2: I don't see the details added ?

---
title: Non graceful shutdown
authors:
  - "@yastij"
owning-sig: sig-storage
participating-sigs:
  - sig-node
approvers:
  - "@saad-ali"
  - "@thockin"
reviewers:
  - "@andrewsykim"
  - "@xing-yang"
editor: Yassine Tijani
creation-date: 2019-06-26
last-updated: 2020-01-16
status: implementable

> ---

Reviewer 3: This can be removed now that there's a kep.yaml file.

Reviewer: Removed.

- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
- [ ] Graduation criteria is in place
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those
approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).

**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
[kubernetes/website]: https://github.com/kubernetes/website

## Summary

> In case of an instance being shutdown (i.e. not running) the control plane doesn't make the right

Reviewer: Done

assumptions to enable stateful workload to fail-over, This KEP introduces a flow to do so.

This KEP is part of collection of KEPs that intends to enhance stateful workload management on top of Kubernetes.

## Motivation

> Today kubernetes doesn't enable providers that are able to safely detach (i.e. only detach when volumes aren't being written to) to do so. This proposal aims to introduce a new CSI capability that if advertised would trigger volume detach if a node is partitioned from the control plane.

Reviewer: Done

### Goals

* Increase the availability of stateful workloads

> * Automate self-healing for stateful workloads

Reviewer 4: Another goal is to target only CSI volumes. A non-goal is the existing in-tree volumes.

Reviewer: Done

### Non-Goals

* Implement in-cluster logic to handle node/control plane partitioning 
* enable detach for all the storage providers

## Proposal

User stories:

> * As a cluster administrator I want my stateful workload to failover in case of a node is partitioned without any intervention when it's safe to do so

Reviewer 2: how to define "a node is partitioned"

Reviewer: Changed to "shut down"

* As a developer I would like to rely on a self-healing platform for my stateful workloads

This KEP plans to introduce the following CSI capabilities.

> * In the [CSI spec](https://github.com/container-storage-interface/spec/blob/master/spec.md), introduce `UNPUBLISH_FENCE` controller service capablity and `FORCE_UNPUBLISH` node service capability.

Reviewer 2: what is FORCE_UNPUBLISH node service?
In what cases, controller/node can have this capability?

Reviewer: Added the CSI spec PR.

> * In the [CSI spec](https://github.com/container-storage-interface/spec/blob/master/spec.md), introduce `UNPUBLISH_FENCE` controller service capablity and `FORCE_UNPUBLISH` node service capability.

Reviewer 5: typo: capablity -> capability

* In the [CSIDriverSpec](https://github.com/kubernetes/kubernetes/blob/v1.20.2/pkg/apis/storage/types.go#L266), add a `SafeDetach` boolean to indicate if it is safe to detach.

```go
// CSIDriverSpec is the specification of a CSIDriver.
type CSIDriverSpec struct {
  ...

  // SafeDetach indicates this CSI volume driver is able to perform detach
  // operations, when set to true, the CSI volume driver needs to ensure that 
  // ControllerUnpublishVolume() implementation is checking for detach safety

>   // +optional

Reviewer 2: what does it mean by "checking for detach safety"?

Reviewer 2: so driver has the capability to detect whether detach is safe to perform? 
Is there an example of such driver and how this driver can make sure detach is safe (validating there is no mount pointing to the disk to be detached?)

  SafeDetach *bool
}

```

> When a node is shut down, the health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to False or Unknown (unreachable) if lease is not renewed for a specific grace period. Node Status becomes NotReady.

Reviewer 4: It would be good to note which parts are existing behavior, and which parts are being added by this proposal.

Reviewer: Done

After 300 seconds (default), the Taint Manager tries to delete Pods on the Node after detecting that the Node is NotReady. The Pods will be stuck in terminating status.

The Pod GC Controller, part of the kube-controller-manager, would need to go through all the Pods in terminating state, verify that the Node is NotReady, and then check if the `SafeDetach` flag in CSIDriver is set to true. If so, forcefully delete pods that are stuck.

Specifically the Pod GC Controller (the only controller able to forcefully delete pods) would also check now for stateful pods, and select those that satisfy ALL of the following conditions:

- pods that are backed by a provisioner that support safe detach
- pods that are backed by volumes that aren't marked as `ReadWriteMany, this would avoid breaking having multiple replicas of the same pod.

> Once pods are selected and forcefully deleted, the attachdetach reconciler should check if the `UNPUBLISH_FENCE` CSI controller capability is true. If true, it should skip the `attachedVolume.MountedByNode` check for the volumes backing these pods and allow `ControllerUnpublishVolume` to happen before `NodeUnpublishVolume` and/or `NodeUnstageVolume` are called.

Reviewer: You are right.  Fixing it.

> This would trigger the deletion of the `volumeAttachement` objects, which would in turn trigger `ControllerUnpublishVolume()`. The provider would need to ensure there that no volume is being used through a mount point. If unable to determine the usage, the storage provider can return an error without detaching the volume, the `external-attacher` would then retry with backoff.

Reviewer 2: could provider be able to check "volume is being used through a mount point."?

Reviewer 4: What are some examples of ways plugins can determine it's safe to detach?

> This would trigger the deletion of the `volumeAttachement` objects, which would in turn trigger `ControllerUnpublishVolume()`. The provider would need to ensure there that no volume is being used through a mount point. If unable to determine the usage, the storage provider can return an error without detaching the volume, the `external-attacher` would then retry with backoff.

Reviewer: It is ControllerUnpublishVolume with a new boolean flag "fence".

> The Pod GC Controller would also apply a `quarantine` taint on the nodes whose pods are forcefully deleted due the shutdown.

Reviewer 5: @yastij wrote below on 18 Jul 2019 below:

> This would be part of a collection of 3 KEPs:
> 1 - node failover (this one)
> 2 - node GC
> 3 - statefulset evictions using NoExecute

why apply a `quarantie` taint instead of `NoExecute` taint?

Reviewer: Can you provide a link to the 3rd KEP "statefulset evictions using NoExecute"?

Reviewer: `NoExecute` is to make sure scheduler does not schedule additional pods to the node. This new taint `quarantine` is to differentiate it from the regular "do not schedule" case.  It indicates the node also needs to be "cleaned up" before new pods can be scheduled on it.

Reviewer 5: there is no  3rd KEP "statefulset evictions using NoExecute" yet. This is mentioned https://github.com/kubernetes/enhancements/pull/1116#issuecomment-512840909

Reviewer 5: We need to create KEP 2, 3 as soon as possible?

> The Pod GC Controller would also apply a `quarantine` taint on the nodes whose pods are forcefully deleted due to the shutdown.

Reviewer 2: what this "quarantine" will take effect?

Reviewer: "quarantine" means pods should not be scheduled on the node.

> ### Handle the Return of Shutdown Node

Reviewer 4: How does this work if the node wasn't shutdown, and was just temporarily partitioned? The pods and containers could still be running but now the volumes are gone Will it get IO errors?

Reviewer: I think we'll handle temporarily partitioned node the same way as shutdown node.  Since the `quarantine` taint is applied on the node, we'll clean up everything on the node when its network connectivity is resumed before removing the taint and allow pods to be scheduled on the node again. Added a note at the end of this section.

Reviewer 5: in order to explain cases not only shutdown but also temporarily partitioned, how about fix this title like:
```
### Handle the Return of Shutdown Node
```
to 
```
### Handle the Return of Partitioned Node
```

If the nodes that were previously shutdown with pods forcefully deleted come back, we need to make sure they are cleaned up before pods are scheduled on them again.

The health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to True if lease is renewed again. Node Status becomes Ready.

> When the node comes up, Kubelet volume manager reconciler will try to clean up mounts by calling `NodeUnpublishVolume`. Since there isn't enough information to reconstruct global mount, Kubelet cannot call `NodeUnstageVolume`. Kubelet volume manager should forcefully cleanup the [global mount directory](/var/lib/kubelet/plugins/kubernetes.io/csi/pv/{pvname}/globalmount) to avoid problems when new pods are scheduled to this node again.

Reviewer: TODO: It may be possible to reconstruct global mount for CSI volumes.

> When the node comes up, Kubelet volume manager reconciler will try to clean up mounts by calling `NodeUnpublishVolume`. Kubelet volume manager reconciler will also try to reconstruct global mount for CSI volumes and call `NodeUnstageVolume`. If there isn't enough information to reconstruct global mount, Kubelet volume manager should forcefully cleanup the [global mount directory](/var/lib/kubelet/plugins/kubernetes.io/csi/pv/{pvname}/globalmount) to avoid problems when new pods are scheduled to this node again.

Reviewer 4: Will kubelet prevent pods from starting too?

> When the node comes up, Kubelet volume manager reconciler will try to clean up mounts by calling `NodeUnpublishVolume`. Kubelet volume manager reconciler will also try to reconstruct global mount for CSI volumes and call `NodeUnstageVolume`. If there isn't enough information to reconstruct global mount, Kubelet volume manager should forcefully cleanup the [global mount directory](/var/lib/kubelet/plugins/kubernetes.io/csi/pv/{pvname}/globalmount) to avoid problems when new pods are scheduled to this node again.

Reviewer 4: What does "forcefully cleanup" mean?

Reviewer: "forcefully cleanup" means deleting those directory without calling "NodeUnstageVolume".  I'll rephrase it.

> The Pod GC Controller should also remove the `quarantine` taint from the nodes that are already cleaned up and are safe to run pods again.

Reviewer 2: one race should avoid by using the quarantine taint is only after detach finished, node can come back to run pods again, I think

Reviewer: Sure.

### Risks and Mitigations

This KEP introduces changes, and if not tested correctly could result in data corruption for users.
To mitigate this we plan to have a high test coverage and to introduce this enhancement as an alpha

## Design Details

### Test Plan

This feature will be tested with the following approaches:

- unit tests
- integration
- e2e

> 

Reviewer 6: This is not a great test plan, please expand on each test to add details, thanks!

we also plan to test this with different version Skews.

### Graduation Criteria

This KEP will be treated as a new feature, and will be introduced with a new feature gate, `NonGracefulFailover`.

This enhancement will go through the following maturity levels: alpha, beta and stable.

Graduation criteria between these levels to be determined.

### Upgrade / Downgrade Strategy

> once enabled by default, the upgrades should be smooth as the default behaviour is off by default. when downgrading/rollbacking with this behaviour enabled, there should be no issues as the control plane with the logic is still there.

Reviewer 4: Also need to expand the upgrade/downgrade story for drivers that will opt-in to this.

Reviewer 7: The current design says there is no code change for storage drivers, so we don't need it, IMO.

- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those
approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).

**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
[kubernetes/website]: https://github.com/kubernetes/website

## Summary

In case of a node being shutdown (i.e. not running) the control plane doesn't make the right
assumptions to enable stateful workload to fail-over. This KEP introduces a flow to do so.

## Motivation

> Today kubernetes doesn't enable storage providers that are able to safely detach volumes (i.e. only detach when volumes aren't being written to) to do so. This proposal aims to introduce a new option in CSIDriver that if set would trigger volume detach if a node is shutdown.

Reviewer 8: > This proposal aims to introduce a new option in CSIDriver that if set would trigger volume detach if a node is shutdown.

It describes how, not what. So, should we move this to `Proposal` section and describe what here?

### Goals

> * Increase the availability of stateful workloads when a node is shutdown.

Reviewer 8: Does it mean that node failure like kernel panic and node slowdown is excluded from the scope and it only handles the case that the node really shutdown properly?

Reviewer: Yes.  Those can be added in a followup KEP after we've reached consensus on a KEP with a narrower scope.

Existing logic:
* When a node is partitioned from the control plane, the health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to False or Unknown (unreachable) if lease is not renewed for a specific grace period. Node Status becomes NotReady.

* After 300 seconds (default), the Taint Manager tries to delete Pods on the Node after detecting that the Node is NotReady. The Pods will be stuck in terminating status.

Proposed logic change:
* The Pod GC Controller, part of the kube-controller-manager, would need to go through all the Pods in terminating state, verify that the Node is NotReady, check if the `Reason` in the `PodStatus` is set to the `NodeShutDownReason` constant, and then check if the `SafeDetach` flag in CSIDriver is set to true. If so, forcefully delete pods that are stuck.

* Specifically the Pod GC Controller (the only controller able to forcefully delete pods) would also check now for stateful pods, and select those that satisfy ALL of the following conditions:

  * pods that are backed by a CSI driver that supports `SafeDetach`.
  * pods that are backed by volumes that aren't marked as `ReadWriteMany`.

> 

Reviewer 5: why `ReadWriteMany` is forbidden?

Reviewer 8: I think that we can already attach a `ReadWriteMany` volume to another pods in other nodes, even when the node that the volume is used on is unresponsive. Therefore, there will be no need for extra special handlings in this case.

Reviewer 4: I'm not sure why we need to explicitly treat RWX specially if a driver says it can support safe detach?

Reviewer: Removed this line.

- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those
approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).

**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
[kubernetes/website]: https://github.com/kubernetes/website

## Summary

In case of a node being shutdown (i.e. not running) the control plane doesn't make the right
assumptions to enable stateful workload to fail-over. This KEP introduces a flow to do so.

> This KEP is part of collection of KEPs that intends to enhance stateful workload management on top of Kubernetes.

Reviewer 9: is this "collection of KEPs" tagged in some way that I can easily track them? wondering how to get a comprehensive overview of proposed changes for stateful workload support.

Reviewer: There's this issue here: https://github.com/kubernetes/enhancements/issues/1403
I'm not aware of a collection of KEPs though.  I'll remove this line.

* In the [CSIDriverSpec](https://github.com/kubernetes/kubernetes/blob/v1.20.2/pkg/apis/storage/types.go#L266), add a `SafeDetach` boolean to indicate if it is safe to detach.

```go
// CSIDriverSpec is the specification of a CSIDriver.
type CSIDriverSpec struct {
  ...

  // SafeDetach indicates this CSI volume driver is able to perform detach
  // operations. When set to true, the CSI volume driver needs to ensure that 
  // ControllerUnpublishVolume() implementation is checking whether a volume
  // can be detached safely
  // +optional
  SafeDetach *bool
}

```

Existing logic:
* When a node is partitioned from the control plane, the health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to False or Unknown (unreachable) if lease is not renewed for a specific grace period. Node Status becomes NotReady.

* After 300 seconds (default), the Taint Manager tries to delete Pods on the Node after detecting that the Node is NotReady. The Pods will be stuck in terminating status.

> Proposed logic change:

Reviewer 9: I think that this flow places too much burden on CSI plugins. The CO knows that the node has problems. The next decision is if/when to "shoot the node", or to allow/wait for it to recover. If waiting for it to recover, the workloads should not be harmed and no state should be migrated. When the waiting game is over, it's time to wreck the node.

*Dismantling the node vs. "shooting" it*

Dismantling the node by making a ton of CSI calls, `O(n)` where `n` is the number of vols, is likely error prone and doesn't scale well (if there are a large number of vols on a node). If multiple nodes are to be dismantled, then it's an `O(m * n)` problem, `m` is number of nodes. Which is even worse. Multiply that by the number of CSI plugins on the node (that support fencing). And also consider that CSI is probably not the only "service provider" that may be interested in this kind of event - who else might want to be part of this flow?

Dismantling the node, as proposed here, also means that more taints are added and pods are evicted. Does eviction apply to *all* pods (even daemonset and/or static pods)? Is it possible to write a pod spec that ignores the quarantine state (if not, should it be)? If pods were using actual hardware/firmware on the host, how does k8s reliably clear dirty state? Other HA systems use power-cycling to clear such state - there's no concept of that in k8s node lifecycle though, right? The WIP CSI PR was trying to promise something along those lines, but w/o the authority to e.g. power cycle (or whatever the infra runtime equivalent is for some cluster) a node (because it's outside the scope of CSI) the PR is a bit hamstrung.

Shooting the node by power cycling it is brutal. And effective w/ respect to dirty firmware/hardware state. Maybe part of the process can still include tainting the node (like the "quarantine" proposed here), and attempting to signal workloads before the node goes down. Though I'd like to see more discussion about what's allowed to run vs. not in that case. I'd also like to see k8s get an API (if it doesn't already have one) that indicates "hey, I've just rebooted and am ready for workloads again - you can trust that there's no dirty state hiding out somewhere waiting to corrupt your workloads"; otherwise, how does k8s know that it's safe to restart workloads on a node that has re-emerged?

Shooting the node also requires k8s to deal w/ "death match" and other fun (read: important) things. See https://ourobengr.com/ha/

Fencing like this should not be the job of individual components (e.g. CSI drivers, generic device drivers, etc.) but instead should be handled higher in the k8s stack. And maybe even involves integration w/ cluster-management APIs (such as Cluster API, owned by SIG Cluster Lifecycle) so that e.g. `Machine` objects are "power cycled" (if they support it), or simply deleted (and then a Cluster API controller can bring up a replacement).

Reviewer: I'm going to narrow down the scope of this KEP and only handle the case when the node is really shutdown, leveraging the work from this KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2000-graceful-node-shutdown

* The Pod GC Controller, part of the kube-controller-manager, would need to go through all the Pods in terminating state, verify that the Node is NotReady, and then check if the `SafeDetach` flag in CSIDriver is set to true. If so, forcefully delete pods that are stuck.

* Specifically the Pod GC Controller (the only controller able to forcefully delete pods) would also check now for stateful pods, and select those that satisfy ALL of the following conditions:

  * pods that are backed by a CSI driver that supports `SafeDetach`.

>   * pods that are backed by volumes that aren't marked as `ReadWriteMany`. This would avoid breaking having multiple replicas of the same pod.

Reviewer 9: "breaking having"? what is this trying to say?

Reviewer: That's a typo.  Will fix it.

Reviewer: Fixed.

* Once pods are selected and forcefully deleted, the attachdetach reconciler should check if the `SafeDetach` flag in CSIDriver is set to true. If so, it should skip the `attachedVolume.MountedByNode` check for the volumes backing these pods and allow `volumeAttachement` to be deleted.

> * This would trigger the deletion of the `volumeAttachement` objects. This would allow `ControllerUnpublishVolume` to happen before `NodeUnpublishVolume` and/or `NodeUnstageVolume` are called.

Reviewer 10: Just want to understand, if this is new behavior or existing logic, with the existing logic as proposed at L110, after the pods get stuck in the terminating state , if we manually delete the pods/volume attachments(assuming safe to delete) , still would go through same workflow right? Issue `ControllerUnpublishVolume` to happen before `NodeUnpublishVolume` and/or `NodeUnstageVolume` are called` and when node comes back up it would issue NodeUnpublishVolume / NodeUnstageVolume as proposed in L134 on checking pods are not running on current node, even with existing logic . Is my understanding correct?

Reviewer: You are right.  If you manually delete pod/va, line 122 should still happen.

Reviewer 9: Intentionally (read: by-design) invoking ControllerUnpublishVolume before NodeUnpublishVolume/NodeUnstageVolume is a violation of the spec:

* https://github.com/container-storage-interface/spec/blob/master/spec.md#controllerunpublishvolume
* https://github.com/container-storage-interface/spec/blob/master/spec.md#nodeunstagevolume

Is there a plan to loosen the CSI spec to permit this? Does this functionality depend on the CSI PR that Ben has been working on re: fencing?

Reviewer 5: Related to what @ramineni said.
when we manually delete pod/va object and then try to issue `NodeUnpublishVolume` / `NodeUnstageVolume` , even if `volumeAttachment` has been deleted `NodeUnpublishVolume` / `NodeUnstageVolume` work correctly? In short, linking information between volume and node is deleted by deleting `volumeAttachment` . Is there no problem to issue `NodeUnpublishVolume` / `NodeUnstageVolume` in this case?

When the node comes up, Kubelet volume manager reconciler will try to clean up mounts by calling `NodeUnpublishVolume`. Kubelet volume manager reconciler will also try to reconstruct global mount for CSI volumes and call `NodeUnstageVolume`. If there isn't enough information to reconstruct global mount, Kubelet volume manager should cleanup the [global mount directory](/var/lib/kubelet/plugins/kubernetes.io/csi/pv/{pvname}/globalmount) to avoid problems when new pods are scheduled to this node again.

The Pod GC Controller should also remove the `quarantine` taint from the nodes that are already cleaned up and are safe to run pods again. We need to make sure the taint is only removed after detach is finished.

Note: We'll handle the temporarily partitioned node the same way as the shutdown node. Since the `quarantine` taint is applied on the node, we'll clean up everything on the node when its network connectivity is resumed before removing the taint and allowing pods to be scheduled on the node again.

### Risks and Mitigations

> This KEP introduces changes, and if not tested correctly could result in data corruption for users.

Reviewer 9: if i'm a user running a stateful workload, how do I opt-out of the new behavior proposed by this KEP? in other words, i'd rather have an administrator deal w/ the fallout resulting from a sick/stuck node vs. k8s possibly getting into a "node crash-loop" state.

e.g. the node enters a non-ready state for some reason unrelated to my stateful workload, which continues to run as normal. it's a critical application and i don't want k8s automatically taking any steps that could result in data loss. where's the policy control for this?

Reviewer: The `SafeDetach` flag in `CSIDriver` should be set to `false` when deploying the CSI driver if the user does not want this behavior.

Reviewer 9: > The `SafeDetach` flag in `CSIDriver` should be set to `false` when deploying the CSI driver if the user does not want this behavior.

OK, I guess I was wondering if it would (or should?) be possible to signal intent on a per-workload basis. It sounds like the answer is "no, it's configured for the cluster as a whole."

Reviewer: Maybe we could consider adding a policy in the StatefulSet. 
https://github.com/kubernetes/kubernetes/blob/d849d9d057369121fc43aa3359059471e1ca9d1c/pkg/apis/apps/types.go#L50

Reviewer: This KEP only handles a real node shutdown case now.

- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
- [ ] Graduation criteria is in place
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those
approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).

**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
[kubernetes/website]: https://github.com/kubernetes/website

## Summary

> In this KEP, we are proposing a way that allows stateful workloads to failover to a different node successfully after the original node is shutdown.

Reviewer 2: I am thinking besides shutdown, if there is hardware failure, or OS broken (any failures that is considered permanent might be addressed in this proposal. But we can mention focus is Shutdown case.

Reviewer: Updated.

## Motivation

The Graceful Node Shutdown [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2000-graceful-node-shutdown) introduced a way to detect a node shutdown and handle it gracefully. However, a node shutdown action may not be detected by Kubelet's Node Shutdown Mananger, either because the command does not trigger the inhibitor locks mechanism used by Kubelet or because of a user error, i.e., the ShutdownGracePeriod and ShutdownGracePeriodCriticalPods are not configured properly.

In this KEP, we are proposing a way to handle node shutdown cases that are not detected by Kubelet's Node Shutdown Manager.

### Use Cases

> * User wants a stateful workload to failover to another running node in case a node is shutdown without any intervention when it's safe to do so.

Reviewer 2: in new proposal, we require user/admin to add taint, so maybe we can say with minimum intervention

Reviewer: Sure.

### Goals

* Increase the availability of stateful workloads when a node is shutdown.
* Automate self-healing for stateful workloads.

### Non-Goals

* Node/control plane partitioning other than a node shutdown is not covered by the proposal, but will be addressed in the future and built on top of this design.
  * We do not have a way to determine whether it is safe to detach in the case of node partitioning. In the Alternatives section, we discussed the node fencing approach and explained why that is not selected for this KEP but will be considered after this KEP is implemented.
* Implement in-cluster logic to handle node/control plane partitioning.

## Proposal

In this section, we are describing the proposed changes to enable workloads to failover to a running node when the original node is shutdown but not detected by Kubelet's Node Shutdown Manager.

Existing logic:

> 1. When a node is partitioned from the control plane, the health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to False or Unknown (unreachable) if lease is not renewed for a specific grace period. Node Status becomes NotReady.

Reviewer 2: "When a node is partitioned from the control plane" -- > when a node is shutdown?

Reviewer: Changed to "when a node is not reachable from the control plane".

> 1. After 300 seconds (default), the Taint Manager tries to delete Pods on the Node after detecting that the Node is NotReady. The Pods will be stuck in terminating status.

Reviewer 2: delete Pods in the graceful way (requires kubelet response to finally delete pod from API server). If kubelet is not responding, pod will be stuck in terminating state.

Reviewer: Do you mean this case will also be handled by the logic proposed in this KEP because Pod will all be stuck in terminating status?

Reviewer 2: oh, I just think to add more details about this point.

Reviewer: Sure.

Proposed logic change:
1. [Proposed change] In the Pod GC Controller, part of the kube-controller-manager, add a new function called gcTerminating. This function would need to go through all the Pods in terminating state, verify that the Node is NotReady. If so, do the following:

>   1. [Proposed change] When a user is sure that a node is shutdown, it will apply a `quarantine` taint on the node.

Reviewer 2: maybe bring up the user action at the beginning. The first step is that user need to confirm node actual state, make sure it is shutdown, or in a non-recoverable state.

Reviewer: Updated.

>   1.  Upon seeing the `quarantine` taint, the Pod GC Controller will forcefully delete the pods on the node. This should happen before the pods are being evicted. The `quarantine` taint indicates that new pods should not be scheduled on the node.

Reviewer 2: "This should happen before the pods are being evicted.

this might not be true, depending when user apply the taint.

Reviewer 2: to avoid user mistakenly put quarantine taint, I think controller could also check node status, at least the node is no longer in ready state, then controller will take actions including deleting pods etc.

Reviewer: The check for node status is there.  I'll add to the KEP.

> This should happen before the pods are being evicted.
> this might not be true, depending when user apply the taint

The logic to forcefully delete the pods are added by this KEP, so we can add logic to check the taint before doing that.

Reviewer 2: If user noticed shutdown and added taint late (more than 5 mins later), the pods are already evicted, GC controller can still force delete the pods, right? That why I think pod force deletion might happen after eviction?

Reviewer 5: Only when node has `quarantine` taint, pod will be forcefully deleted by the Pod GC Controller, I guess. 
```
Upon seeing the `quarantine` taint, the Pod GC Controller will forcefully delete the pods on the node. 
```

>     1. new `quarantine` taint: no execute effect, meaning it will evict pods and start on different node.

Reviewer 2: I think quarantine taint might use NoSchedule effect. Not sure NoExecute will also means NoSchedule or not.

Taint example, 
quarantine=nodeshutdown:NoSchedule
quarantine=hardwarefailure:NoSchedule

Reviewer: Done

1. [Proposed change] Once pods are selected and forcefully deleted, the attachdetach reconciler should check the `quarantine` taint on the node. If the taint is present, the attachdetach reconciler will not wait for 6 minutes to do force detach. Instead it will force detach right away and allow `volumeAttachement` to be deleted.

1. This would trigger the deletion of the `volumeAttachement` objects. This would allow `ControllerUnpublishVolume` to happen before `NodeUnpublishVolume` and/or `NodeUnstageVolume` are called.

1. When the `external-attacher` detects the `volumeAttachement` object is being deleted, it calls CSI driver's `ControllerUnpublishVolume`.

> 1. [Proposed change] The CSI driver would need to ensure that no volume is being used through a mount point. If unable to determine the usage, the CSI driver can return an error without detaching the volume. The `external-attacher` would then retry with backoffs.

Reviewer 2: This might not necessary considering some driver (controller side) cannot verify mountpoint is being used. 

by adding quarantine taint, we are certain that no mount points should exist and being used?

Reviewer: This was from previous version of the KEP. I'll remove this.

### Handle the Return of Shutdown Node

If the nodes that were previously shutdown with pods forcefully deleted come back, we need to make sure they are cleaned up before pods are scheduled on them again.

The health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to True if lease is renewed again. Node Status becomes Ready.

> When the node comes up, Kubelet volume manager reconciler will try to clean up mounts by calling `NodeUnpublishVolume`.

Reviewer 2: I thought when node is shutdown, no mountpoint should exist any more. Do we need this part?

Reviewer: This is existing code in the reconciler. I'll remove from the KEP.

> [Proposed Change] Pod GC Controller automatically removes the `quarantine` taint. It checks to make sure no pods are attached to a Node, and no volume attachments on Node. If so, it removes the taint. This makes sure no new pods are scheduled on the node before the `quarantine` taint is removed.

Reviewer 2: maybe move this part to the previous section or at the beginning this section?

Reviewer: Moved up a little bit.

  * We do not have a way to determine whether it is safe to detach in the case of node partitioning. In the Alternatives section, we discussed the node fencing approach and explained why that is not selected for this KEP but will be considered after this KEP is implemented.
* Implement in-cluster logic to handle node/control plane partitioning.

## Proposal

In this section, we are describing the proposed changes to enable workloads to failover to a running node when the original node is shutdown but not detected by Kubelet's Node Shutdown Manager, or when the node is in other non-recoverable cases such as the hardware failure or broken OS.

Existing logic:
1. When a node is not reachable from the control plane, the health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to False or Unknown (unreachable) if lease is not renewed for a specific grace period. Node Status becomes NotReady.

1. After 300 seconds (default), the Taint Manager tries to delete Pods on the Node after detecting that the Node is NotReady. The Pods will be stuck in terminating status.

Proposed logic change:

> 1. [Proposed change] This proposal requires a user to apply a `quarantine` taint on a node when the user has confirmed that this node is shutdown or in a non-recoerable state due to the hardware failure or broken OS.

Reviewer 11: This says "user" - does it mean "or controller" ?

Reviewer 11: Why is it a taint?  Taints imply that a toleration can overcome it, but if it is off...

Reviewer 11: Reading this it feels more like a `node.status.powerState` field than a taint.  E.g. `powerState = ConfirmedOff".

The scheduler would not select nodes that are not Ready, which would necessarily include powered-off.

Cloud controllers would be responsible for poking at NotReady nodes and seeing if they can be confirmed as off. (and a logical next step would be to turn them off after X time, when possible -> fencing).

Reviewer 2: In the previous design, we are thinking to allow either user manually check machine is off or cloud provider automatically check machine status is shutdown. But there is a concern of checking automatically by cloud provider, whether it can differentiate between node is shutdown or it is in the middle of rebooting. So in this current design, we want to only allow user to manually check the status and make sure machine is off before adding the taint.

Yes, the taint here is mainly a signal about the situation so that controller can take actions based on this signal. It is not necessary to use taint, but it is easy to use (allowing either user or controller could update it) in this situation. It is a good point that in case there is some tolerations setting which might overwrite the logic here. Annotation is another option here, but it seems not recommended due to its limitations?

In this KEP we want to reduce the scope so that we can make some progress for this alpha feature. There are some other user cases that we plan to address in the future, that could build upon on this, e.g., fencing, putting machine in maintenance mode explicitly for hardware upgrade etc.

Reviewer 2: In this proposal, we propose to use "NoExecute" effect for quarantine taint of shutdown node. so I think there are two options about pod tolerations.
1. follow taint and toleration policy: If pod is set to tolerate all taints and effects, that means user does NOT want to evict pods when node is not ready. So GC controller will filter out those pods and only force delete pods that do not have a matching toleration.
2. ignore toleration setting: GC controller delete ALL pods that are scheduled on the shutdown node since there is point of letting pods stay in the node that is already shutdown. But this will break the taint and toleration policy.
So I think option 1 is preferable considering the taint and toleration policy.

Reviewer 2: In this proposal, we propose to use "NoExecute" effect for quarantine taint of shutdown node. so I think there are two options about pod tolerations.
1. follow taint and toleration policy: If pod is set to tolerate all taints and effects, that means user does NOT want to evict pods when node is not ready. So GC controller will filter out those pods and only force delete pods that do not have a matching toleration.
2. ignore toleration setting: GC controller delete ALL pods that are scheduled on the shutdown node since there is point of letting pods stay in the node that is already shutdown. But this will break the taint and toleration policy.
So I think option 1 is preferable considering the taint and toleration policy.

Reviewer: Updated KEP with option 1.

1. [Proposed change] In the Pod GC Controller, part of the kube-controller-manager, add a new function called gcTerminating. This function would need to go through all the Pods in terminating state, verify that the Node is NotReady. If so, do the following:
  1.  Upon seeing the `quarantine` taint, the Pod GC Controller will forcefully delete the pods on the node. This should happen before the pods are being evicted. The `quarantine` taint indicates that new pods should not be scheduled on the node.
  1. new `quarantine` taint has NoSchedule effect, meaning it will not schedule on the shutdown node. For example, `quarantine=nodeshutdown:NoSchedule` or `quarantine=hardwarefailure:NoSchedule`.

1. [Proposed change] Once pods are selected and forcefully deleted, the attachdetach reconciler should check the `quarantine` taint on the node. If the taint is present, the attachdetach reconciler will not wait for 6 minutes to do force detach. Instead it will force detach right away and allow `volumeAttachement` to be deleted.

> 

Reviewer 5: attachdetach reconciler itself do force detach? If so, how to do?

Reviewer 2: actually there is no actual difference between a normal detach operation and a "force detach". We call it force here mainly because normally controller will not trigger detach at all if node status showing volume is still in use (means volume is still mounted) by kubelet. But here we are sure that node is in shut down mode, we can trigger detach, no matter whether node status showing volume is in use or not..

Reviewer 5: @jingxu97 Thank you, I understand :)

> 1. This would trigger the deletion of the `volumeAttachement` objects. This would allow `ControllerUnpublishVolume` to happen before `NodeUnpublishVolume` and/or `NodeUnstageVolume` are called.

Reviewer 7: It's out of the current state transition of PVs in CSI spec, right? IMO, it's good to update it **after** this KEP will be accepted, but  we should note it as the impact of this proposal. Otherwise, it looks having no impact to CSI drivers.

Reviewer 7: And also, any note for in-tree storage drivers? The current non-goals contains no limitation for in-tree drivers.

Reviewer: [Proposed change] indicates there's a code change. After the proposed change on line 105, line 107 will happen automatically without additional code change.  I can clarify that in the KEP.

1. When the `external-attacher` detects the `volumeAttachement` object is being deleted, it calls CSI driver's `ControllerUnpublishVolume`.

### Handle the Return of Shutdown Node

If the nodes that were previously shutdown with pods forcefully deleted come back, we need to make sure they are cleaned up before pods are scheduled on them again.

> [Proposed Change] Pod GC Controller automatically removes the `quarantine` taint. It checks to make sure no pods are attached to a Node, and no volume attachments on Node. If so, it removes the taint. This makes sure no new pods are scheduled on the node before the `quarantine` taint is removed.

Reviewer 7: Do we need an option to disable automatic removing the `quarantine` taint, for example, to examine the root cause of node failure?

Reviewer: Changed the KEP to require the user to manually remove the `quarantine` taint. Pod GC Controller can be enhanced in the future to automatically remove it.

  - [Goals](#goals)
  - [Non-Goals](#non-goals)
- [Proposal](#proposal)
  - [Handle the Return of Shutdown Node](#handle-the-return-of-shutdown-node)
  - [Risks and Mitigations](#risks-and-mitigations)
- [Design Details](#design-details)
  - [Test Plan](#test-plan)
  - [Unit tests](#unit-tests)
  - [E2E tests](#e2e-tests)
  - [Graduation Criteria](#graduation-criteria)
    - [Alpha](#alpha)
    - [Alpha -&gt; Beta Graduation](#alpha---beta-graduation)
    - [Beta -&gt; GA Graduation](#beta---ga-graduation)
  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
- [Production Readiness Review Questionnaire](#production-readiness-review-questionnaire)
  - [Feature Enablement and Rollback](#feature-enablement-and-rollback)
  - [Rollout, Upgrade and Rollback Planning](#rollout-upgrade-and-rollback-planning)
  - [Monitoring Requirements](#monitoring-requirements)
  - [Dependencies](#dependencies)
  - [Scalability](#scalability)
  - [Troubleshooting](#troubleshooting)
- [Implementation History](#implementation-history)
- [Alternatives](#alternatives)
  - [Node fencing](#node-fencing)
  - [SafeDetach Option](#safedetach-option)
<!-- /toc -->

## Release Signoff Checklist

> - [] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
> - [ ] KEP approvers have set the KEP status to `implementable`
> - [ ] Design details are appropriately documented
> - [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
> - [ ] Graduation criteria is in place

Reviewer 7: ```suggestion
Items marked with (R) are required *prior to targeting to a milestone / release*.

- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
- [ ] (R) KEP approvers have approved the KEP status as `implementable`
- [ ] (R) Design details are appropriately documented
- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
  - [ ] e2e Tests for all Beta API Operations (endpoints)
  - [ ] (R) Ensure GA e2e tests for meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
```

Reviewer: Done

- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]

> - [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

Reviewer 7: ```suggestion
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
```

1. [Proposed change] Once pods are selected and forcefully deleted, the attachdetach reconciler should check the `quarantine` taint on the node. If the taint is present, the attachdetach reconciler will not wait for 6 minutes to do force detach. Instead it will force detach right away and allow `volumeAttachment` to be deleted.

1. This would trigger the deletion of the `volumeAttachment` objects. This would allow `ControllerUnpublishVolume` to happen before `NodeUnpublishVolume` and/or `NodeUnstageVolume` are called. Note that there is no additional code changes required for this step. This happens automatically after the `Proposed change` in the previous step to force detach right away.

1. When the `external-attacher` detects the `volumeAttachment` object is being deleted, it calls CSI driver's `ControllerUnpublishVolume`.

### Handle the Return of Shutdown Node

If the nodes that were previously shutdown with pods forcefully deleted come back, we need to make sure they are cleaned up before pods are scheduled on them again.

> [Proposed Change] We require the user to manually remove the `quarantine` taint after the pods are moved to a new node since the user was the one who originally added the taint. This makes sure no new pods are scheduled on the node before the `quarantine` taint is removed. In the future, we can enhance the Pod GC Controller to automatically remove the `quarantine` taint, after verifying that no pods are attached to the node, and no volume attachments on the node.

Reviewer 7: How about adding "Note that there is no additional code change required for the return of shutdown node"?

Reviewer: Sure.

**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
[kubernetes/website]: https://github.com/kubernetes/website

## Summary

In this KEP, we are proposing a way that allows stateful workloads to failover to a different node successfully after the original node is shutdown or in a non-recoverable state such as the hardware failure or broken OS.

## Motivation

The Graceful Node Shutdown [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2000-graceful-node-shutdown) introduced a way to detect a node shutdown and handle it gracefully. However, a node shutdown action may not be detected by Kubelet's Node Shutdown Mananger, either because the command does not trigger the inhibitor locks mechanism used by Kubelet or because of a user error, i.e., the ShutdownGracePeriod and ShutdownGracePeriodCriticalPods are not configured properly.

> In this KEP, we are proposing a way to handle node shutdown cases that are not detected by Kubelet's Node Shutdown Manager.

Reviewer 4: What's not clear this section is the impact to workloads in this scenario. It would be good to explain why StatefulSet pods are stuck forever and why we need this new behavior to fix that.

Reviewer: Added why we need this fix in Motivation section.

  * We do not have a way to determine whether it is safe to detach in the case of node partitioning. In the Alternatives section, we discussed the node fencing approach and explained why that is not selected for this KEP but will be considered after this KEP is implemented.
* Implement in-cluster logic to handle node/control plane partitioning.

## Proposal

In this section, we are describing the proposed changes to enable workloads to failover to a running node when the original node is shutdown but not detected by Kubelet's Node Shutdown Manager, or when the node is in other non-recoverable cases such as the hardware failure or broken OS.

Existing logic:
1. When a node is not reachable from the control plane, the health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to False or Unknown (unreachable) if lease is not renewed for a specific grace period. Node Status becomes NotReady.

1. After 300 seconds (default), the Taint Manager tries to delete Pods on the Node after detecting that the Node is NotReady. The Pods will be stuck in terminating status.

Proposed logic change:

> 1. [Proposed change] This proposal requires a user to apply a `quarantine` taint on a node when the user has confirmed that this node is shutdown or in a non-recoerable state due to the hardware failure or broken OS.

Reviewer 2: maybe add note to make sure user only add this taint if node is not coming back at least for some time. If node is in the middle of restarting, they should not use this taint.

> 1. [Proposed change] In the Pod GC Controller, part of the kube-controller-manager, add a new function called gcTerminating. This function would need to go through all the Pods in terminating state, verify that the Node is NotReady. If so, do the following:

Reviewer 4: How does this interact with the Taint Manager specified earlier?

Reviewer: The Taint Manager mentioned earlier is NoExecuteTaintManager. The message I see in the logs is "NoExecuteTaintManager is deleting pod".

The new taint mentioned here will be TaintEffectNoSchedule so it is not handled by NoExecuteTaintManager.

>   1.  Upon seeing the `quarantine` taint, the Pod GC Controller will forcefully delete the pods on the node. This should happen before the pods are being evicted. The `quarantine` taint indicates that new pods should not be scheduled on the node.

Reviewer 11: +1 - saying "should happen" is ambiguous - does it have to happen?  Is that an input or output condition?

>   1. new `out-of-service` taint has NoSchedule effect, meaning new pod will not schedule on the shutdown node unless it has a matching toleration. For example, `out-of-service=nodeshutdown:NoSchedule` or `out-of-service=hardwarefailure:NoSchedule`.

Reviewer 11: Two thoughts:

1) You're making the user do it - you can specify which effect you want, but you have to define what happens, regardless of what effect the user specifies (i.e. humans make mistakes).

2) Why are you specifying NoSchedule and then implementing it as NoExecin your own controller?  Why not specify it as NoExec?

Would it be simpler to define this as somethign like "the attachdetach reconciler will look for evicted pods on nodes which are tainted as 'out-of-service', and will bypass the usual 6 minute  wait to force detach volumes"?  That way, either taint effect works.

What am I missing?  I feel like there is some state-transition that you are trying to haandle that I am missing.

Reviewer 4: @thockin and I discussed. It seems to make more sense to use NoExecute rather than NoScheule for this behavior. The logic could be an extension of the existing logic that handles NoExecute. As for whether this should be a new taint effect or a special case of NoExecute, we agreed that most users will likely want similar behavior as NoExecute (people who tolerate NoExecute will also want to tolerate force delete. People who don't tolerate NoExec will also not want to tolerate force delete), so it's ok to make this to a special well-known-taint.

Reviewer 11: ...unless I amissing something?

It seems overll simpler to say:
* taints work as normal
* if a pod has a deletionTimestamp *and* the node is out-of-service, it's OK to force delete
* if a VolumeAttachment points to a node that is out-of-service *and* no running pods use it, it's OK to force detach

Reviewer: I changed to NoExecute.  Please take another look.

>   1. We'll follow taint and toleration policy. If a pod is set to tolerate all taints and effects, that means user does NOT want to evict pods when node is not ready. So GC controller will filter out those pods and only forcefully delete pods that do not have a matching toleration.

Reviewer 2: Another choice is "NoExecute" effect. Both effect will not schedule new pods to the shutdown node. The difference is "NoExecute" will also evict pods unless pods tolerate this effect.

>   1. new `quarantine` taint has NoSchedule effect, meaning it will not schedule on the shutdown node. For example, `quarantine=nodeshutdown:NoSchedule` or `quarantine=hardwarefailure:NoSchedule`. Although both `NoSchedule` and ``NoExecute` will not schedule new pods to the shutdown node, the difference is `NoExecute` will also evict pods unless pods tolerate this effect. So `NoSchedule` is chosen in this proposal.

Reviewer 4: When you have a taint with `NoSchedule`, it means that new pods won't be scheduled to the node, but any existing pods on that node will continue to run. It seems like the new behavior being proposed here is not honoring that, and it could actually cause Pods to be forcefully evicted from the node, which is closer to what `NoExecute` does. I think it's better that we create a new taint effect, maybe something like `NoExecuteForce`, to control this new behavior, rather than special casing a taint key with behavior that doesn't match the effect.

>   1. We'll follow taint and toleration policy. If a pod is set to tolerate all taints and effects, that means user does NOT want to evict pods when node is not ready. So GC controller will filter out those pods and only forcefully delete pods that do not have a matching toleration.

Reviewer 7: So, users can also use NoExecute because tolerating all taints and effects will ignore it, right?

Reviewer: Just changed to NoExecute.

1. [Proposed change] Once pods are selected and forcefully deleted, the attachdetach reconciler should check the `quarantine` taint on the node. If the taint is present, the attachdetach reconciler will not wait for 6 minutes to do force detach. Instead it will force detach right away and allow `volumeAttachment` to be deleted.

> 1. This would trigger the deletion of the `volumeAttachment` objects. This would allow `ControllerUnpublishVolume` to happen before `NodeUnpublishVolume` and/or `NodeUnstageVolume` are called. Note that there is no additional code changes required for this step. This happens automatically after the `Proposed change` in the previous step to force detach right away.

Reviewer 2: here the explanation is for CSI case. This KEP should also work for in-tree case.

Reviewer 2: `NodeUnpublishVolume` and/or `NodeUnstageVolume` will not be called since node is shutdown already?

> 1. [Proposed change] Once pods are selected and forcefully deleted, the attachdetach reconciler should check the `quarantine` taint on the node. If the taint is present, the attachdetach reconciler will not wait for 6 minutes to do force detach. Instead it will force detach right away and allow `volumeAttachment` to be deleted.

Reviewer 4: What is the behavior today if we wait 6 minutes? Will detach be called?

1. When the `external-attacher` detects the `volumeAttachment` object is being deleted, it calls CSI driver's `ControllerUnpublishVolume`.

### Handle the Return of Shutdown Node

If the nodes that were previously shutdown with pods forcefully deleted come back, we need to make sure they are cleaned up before pods are scheduled on them again.

> [Proposed Change] We require the user to manually remove the `quarantine` taint after the pods are moved to a new node since the user was the one who originally added the taint. This makes sure no new pods are scheduled on the node before the `quarantine` taint is removed. In the future, we can enhance the Pod GC Controller to automatically remove the `quarantine` taint, after verifying that no pods are attached to the node, and no volume attachments on the node.

Reviewer 4: Has the node rebooted at this point? Can kubelet recover appropriately if it wasn't rebooted?

Reviewer 5: @msau42 Yes, I think your recognition is right.
```suggestion
[Proposed Change] We require the user to manually remove the `quarantine` taint after the pods are moved to a new node and the user checked that node has been recovered since the user was the one who originally added the taint. This makes sure no new pods are scheduled on the node before the `quarantine` taint is removed. In the future, we can enhance the Pod GC Controller to automatically remove the `quarantine` taint, after verifying that no pods are attached to the node, and no volume attachments on the node.
```

Reviewer: Fixed.

Note that there is no additional code change required for the return of shutdown node.

[Existing Logic] The health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to True if lease is renewed again. Node Status becomes Ready.

> [Existing Logic] Note that if the `quarantine` taint was not added to the shutdown node, the pods will be stuck in terminating status forever if the shutdown node remains down. If the shutdown node comes up, however, the original `volumeAttachment` will be deleted and the pods will be moved to a new node successfully.

Reviewer 4: Who deletes the original VolumeAttachment in the existing logic?

Reviewer 5: attach detach controller does.

Similarly, this approach can also be applied to the case when a node is in a non-recoverable state such as the hardware failure or broken OS.

### Use Cases

* If user wants to intentionally shutdown a node, he/she can validate whether graceful node shutdown feature works. If Kubelet is able to detect node is shutting down, it will gracefully delete pods, and new pods will be created on another running node.
* If graceful shutdown is not working or node is in non-recoverable state due to hardware failure or broken OS, etc., user now can enable this feature, and add `out-of-service=nodeshutdown:NoExecute` taint which will be explained in detail below to trigger non-graceful shutdown behavior.

### Goals

* Increase the availability of stateful workloads when a node is shutdown, or in other non-recoverable cases such as the hardware failure or broken OS.
* Automate self-healing for stateful workloads.

> 

Reviewer 2: may add a line in Goals or even Non-goals mentioning "this proposal requires user intervention to detect node shutdown/failure cases, because currently there is no good way of detecting whether node is in the middle of restarting. Once this feature in this KEP is developed and tested, plan to work on automatic way of detecting and fencing node for shutdown/failure use cases, using a different KEP.

Reviewer: Added to the Goals section.

  * Once the feature in this KEP is developed and tested, we plan to work on automatic way of detecting and fencing node for shutdown/failure use cases and re-evaluating some approaches listed in the Alternatives section.

### Non-Goals

* Node/control plane partitioning other than a node shutdown is not covered by the proposal, but will be addressed in the future and built on top of this design.
  * We do not have a way to determine whether it is safe to detach in the case of node partitioning. In the Alternatives section, we discussed the node fencing approach and explained why that is not selected for this KEP but will be considered after this KEP is implemented.
* Implement in-cluster logic to handle node/control plane partitioning.

## Proposal

In this section, we are describing the proposed changes to enable workloads to failover to a running node when the original node is shutdown but not detected by Kubelet's Node Shutdown Manager, or when the node is in other non-recoverable cases such as the hardware failure or broken OS.

> Note: The proposal in the current KEP involves user manually add/remove a taint. In the future, we will look into a more automatic approach that does not require these manual steps.

Reviewer 2: As @derekwaynecarr  suggested here https://github.com/kubernetes/enhancements/pull/1116/files#r797999208, we can make it more explicit about user actions:

When users verify that a node is already in shutdown or power off state (not in the middle of restarting), either user intentionally shut it down or node is down due to hardware failures, OS issues, they can taint the node with "out-of-service" with NoExecute effect.

Reviewer: Added.

Existing logic:
1. When a node is not reachable from the control plane, the health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to False or Unknown (unreachable) if lease is not renewed for a specific grace period. Node Status becomes NotReady.

1. After 300 seconds (default), the Taint Manager tries to delete Pods on the Node after detecting that the Node is NotReady. The Pods will be stuck in terminating status.

Proposed logic change:
1. [Proposed change] This proposal requires a user to apply a `out-of-service` taint on a node when the user has confirmed that this node is shutdown or in a non-recoverable state due to the hardware failure or broken OS. Note that user should only add this taint if the node is not coming back at least for some time. If the node is in the middle of restarting, this taint should not be used.

> 1. [Proposed change] In the Pod GC Controller, part of the kube-controller-manager, add a new function called gcTerminating. This function would need to go through all the Pods in terminating state, verify that the Node is NotReady. If so, do the following:

Reviewer 12: to be clear, perhaps

s/verify that the Node is NotReady/ verify that the node the pod scheduled is on is `NotReady`

Reviewer: Done.

>   1.  Upon seeing the `out-of-service` taint, the Pod GC Controller will forcefully delete the pods on the node if there are no matching tolation on the pods. This new `out-of-service` taint has `NoExecute` effect, meaning the pod will be evicted and a new pod will not schedule on the shutdown node unless it has a matching toleration. For example, `out-of-service=nodeshutdown:NoExecute` or `out-of-service=hardwarefailure:NoExecute`. We suggest to use `NoExecute` effect in taint to make sure pods will be evicted (deleted) and fail over to other nodes.

Reviewer 13: so instead of the admin force deleting pods, the admin is force tainting the node, and deferring force delete to a controller?  i can see that as being a ux benefit.

Reviewer 2: yes, admin just needs to taint the node, all other actions are handled by controllers

Reviewer 14: So this proposal does not uses existing shutdown taint, which we added to handle the case of automatic detection of nodes which were shutdown via cloudprovider APIs - `cloudproviderapi.TaintNodeShutdown,`. Are we planning to remove that taint or reuse it or just leave it?

Reviewer: We are not going to use that taint because it is serving a different purpose for the Cloud Providers.

Reviewer 14: ack. lets revisit this when we do automatic detection of nodes that have been shutdown. AFAIK - existing taint serves no useful purpose currently (but I wonder if some external operators depend on it).

Reviewer: Sounds good.

**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
[kubernetes/website]: https://github.com/kubernetes/website

## Summary

In this KEP, we are proposing a way that allows stateful workloads to failover to a different node successfully after the original node is shutdown or in a non-recoverable state such as the hardware failure or broken OS.

## Motivation

The Graceful Node Shutdown [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2000-graceful-node-shutdown) introduced a way to detect a node shutdown and handle it gracefully. However, a node shutdown action may not be detected by Kubelet's Node Shutdown Mananger, either because the command does not trigger the inhibitor locks mechanism used by Kubelet or because of a user error, i.e., the ShutdownGracePeriod and ShutdownGracePeriodCriticalPods are not configured properly.

> When a node is shutdown but not detected by Kubelet's Node Shutdown Manager, the pods that are part of a StatefulSet will be stuck in terminating status on the shutdown node and cannot move to a new running node. This is because the VolumeAttachments are not deleted from the original shutdown node so the volumes used by these pods cannot be attached to a new running node. As a result, the application running on the StatefulSet cannot function properly. If the original shutdown node comes up, the pods will move to a different running node. If the original shutdown node does not come up, these pods will be stuck in terminating status on the shutdown node forever.

Reviewer 4: I don't think the crux of the issue is caused by VolumeAttachments. If you run a StatefulSet without any volumes, you should see the same issue.

The primary issue is that kubelet is normally responsible for deleting the Pod object, and in the shutdown case, kubelet is not available to do that, so the StatefulSet controller cannot create a new Pod with the same name. This is a permanent state that cannot be resolved without human intervention or kubelet coming back.

The VolumeAttachment issue is not a permanent state. The force detach after 6 min timeout will cleanup the VolumeAttachment eventually.

Reviewer: Fixed.

  * We do not have a way to determine whether it is safe to detach in the case of node partitioning. In the Alternatives section, we discussed the node fencing approach and explained why that is not selected for this KEP but will be considered after this KEP is implemented.
* Implement in-cluster logic to handle node/control plane partitioning.

## Proposal

In this section, we are describing the proposed changes to enable workloads to failover to a running node when the original node is shutdown but not detected by Kubelet's Node Shutdown Manager, or when the node is in other non-recoverable cases such as the hardware failure or broken OS.

Existing logic:
1. When a node is not reachable from the control plane, the health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to False or Unknown (unreachable) if lease is not renewed for a specific grace period. Node Status becomes NotReady.

1. After 300 seconds (default), the Taint Manager tries to delete Pods on the Node after detecting that the Node is NotReady. The Pods will be stuck in terminating status.

Proposed logic change:

> 1. [Proposed change] This proposal requires a user to apply a `out-of-service` taint on a node when the user has confirmed that this node is shutdown or in a non-recoverable state due to the hardware failure or broken OS. Note that user should only add this taint if the node is not coming back at least for some time. If the node is in the middle of restarting, this taint should not be used.

Reviewer 11: I'd add something like: "The user setting the taint may use either the NoSchedule or the NoExecute effect, but since this proposal only affects terminating pods, they probably want NoExecute"

Reviewer: If we add this to the KEP, does that mean in Pod GC Controller, we need to check the new taint which could be either NoSchedule and NoExecute effect?  Should we keep it simple and just use NoExecute?

Reviewer 4: If a user set NoSchedule, then we don't do anything.

Reviewer: I'm not getting this. If we don't do anything, then do we have to ask user to apply a new taint?

Reviewer 2: When node is not ready, the taint with "NoExecute" effect will be added. So we cannot just check the taint effect, need to detect by its key "out-of-service" suggested here.

Reviewer 2: node lifecycle controller automatically add the following node when kubelet is not responding 
node.kubernetes.io/not-ready: NoSchedule (need to confirm)
node.kubernetes.io/unreachable: NoExecute  --> normal eviction (gracefully delete)

Here we propose user to add new taint
node.kubernetes.io/out-of-service: NoExecute --> force delete

But the key name seems not easy to differentiate with unreachable, I feel.

Reviewer 7: ‘unreachable’ node will be back to the service if it is reachable again, but ‘out-of-service’ node won’t (or shouldn’t) because a user confirmed that it was down or had a problem and decided making it out-of-service.

Reviewer 11: The point I am making is this:

There are several loosely-related things that need to happen for this KEP to tak eeffect:

1) stop scheduling to this node - any taint-effect will suffiice

2) evict some or all pods from this node - NoExecute will work, but the user might want a custom controller and why would we disallow that?

3) mark the node as out-of-service - using a taint for this is convenient since it works semantically

So why do we care which effect the out-of-service taint has?  All we care about is a) the node is OOS; b) a pod-with-volume is terminating.  NoExecute achieves both, but ANY mechanism that satisfies (a) and (b) should be sufficient, right?  Am I missing somethign?

Reviewer 2: So if user add a taint "node.kubernetes.io/out-of-service:NoSchedule", what behavior should be expected? By definition, NoSchedule means
```
        // Do not allow new pods to schedule onto the node unless they tolerate the taint,
	// but allow all pods submitted to Kubelet without going through the scheduler
	// to start, and allow all already-running pods to continue running.
	// Enforced by the scheduler.
```

In this case, should controller still go ahead to force delete pods, then the behavior will be different from what is "NoSchedule" is defined?

Or maybe you are proposing the new logic in controller here should not directly trigger deleting pod first. If the OOS taint is added, it will check which pods are in terminating state, and force delete those pods? If somehow no pods are in terminating state, controller will not do anything?

Reviewer 11: > Or maybe you are proposing the new logic in controller here should not directly trigger deleting pod first.

Exactly.  The magic of this KEP is that we get new information - "out-of-service", which we can intersect with "terminating" to do smarter / faster cleanup.

> If somehow no pods are in terminating state, controller will not do anything?

That sounds correct to me - maybe the user has some other process which will delete those pods in 30 seconds, so they set NoSchedule and let their own controller do it's thing.  That will eventually delete the pods, and the new (this KEP) logic kicks in.

Does that make sense?

Reviewer 2: Got it. I think the above logic does provide more control knobs and flexibility. We need some document/guide for users to understand when they are enable this feature, what are the behaviors to expect in different situations.

Reviewer 13: Is there any guidance for when a user should apply this taint?  I am trying to think about what automated system would apply this taint.

Reviewer: In the "Proposed logic change" section, we mentioned that "Note that user should only add this taint if the node is not coming back at least for some time. If the node is in the middle of restarting, this taint should not be used. "  An automated system can probably check how long a node has been NotReady.

Reviewer 7: IMO,
* Before a user applies `out-of-service` taint to a shutdown node or a failed node, a user should confirm that it is powered off. If not, a user should make it powered off.
* Before a user turns on the power of a node with `out-of-service` taint, a user should confirm that there is no Pod and no attached PV on it with kubectl or so, even a user won’t make it returned a cluster.
* Before a user removes `out-of-service` taint from a node to restart hosting Pods, a user should confirm that:
    * its status is Ready with kubectl or so
    * there is no attached extra storage for PV on it with SSH or so

> 1. [Proposed change] In the Pod GC Controller, part of the kube-controller-manager, add a new function called gcTerminating. This function would need to go through all the Pods in terminating state, verify that the Node is NotReady. If so, do the following:

Reviewer: Yes, NoExecuteTaintManager tries to delete Pods already in existing code as there's already a NoExecute taint on the Pod when node becomes Not Ready. The TaintManager will try to delete the Pod if there is "at least" one NoExecute taint.  So adding another out-of-service NoExecute taint will not change this behavior. We do not need to change the code in TaintManager.

The Pods will be stuck in terminating status as described in the KEP. They will be forcefully deleted by the Pod GC controller as proposed in the KEP.

> 1. [Proposed change] In the Pod GC Controller, part of the kube-controller-manager, add a new function called gcTerminating. This function would need to go through all the Pods in terminating state, verify that the Node is NotReady. If so, do the following:

Reviewer 4: I think it may be cleaner to have all logic handling NoExecute taint to be in the same controller, wdyt?

> 1. [Proposed change] In the Pod GC Controller, part of the kube-controller-manager, add a new function called gcTerminating. This function would need to go through all the Pods in terminating state, verify that the Node is NotReady. If so, do the following:

Reviewer: Pod GC Controller is the only controller that forcefully deletes pods (grace period = 0). This is probably on purpose because it does garbage collection.  I'm not sure if we should let NoExecuteTaintManager also forcefully deletes pods.

> 1. [Proposed change] In the Pod GC Controller, part of the kube-controller-manager, add a new function called gcTerminating. This function would need to go through all the Pods in terminating state, verify that the Node is NotReady. If so, do the following:

Reviewer 7: [Taints and Tolerations](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/#concepts) says:

> Normally, if a taint with effect NoExecute is added to a node, then any pods that do not tolerate the taint will be evicted **immediately**, and pods that do tolerate the taint will never be evicted.

So, Taint Manager also deletes pods forcefully if it meant no grace period.

**But**, I'm ok that Pod GC Controller will do it because stalled terminating pods are just garbage.

> 1. [Proposed change] In the Pod GC Controller, part of the kube-controller-manager, add a new function called gcTerminating. This function would need to go through all the Pods in terminating state, verify that the Node is NotReady. If so, do the following:

Reviewer: In Pod GC Controller, it uses this NewDeleteOptions{0}. That means grace period is set to 0.
https://github.com/kubernetes/kubernetes/blob/v1.24.0-alpha.1/pkg/controller/podgc/gc_controller.go#L79

In TaintManager, it uses DeleteOptions{}. That means default grace period.
https://github.com/kubernetes/kubernetes/blob/v1.24.0-alpha.1/pkg/controller/nodelifecycle/scheduler/taint_manager.go#L112

I searched for NewDeleteOptions{0} and only found it is used in Pod GC Controller. I also don't see DeleteOptions with GracePeriodSeconds set to 0 directly.
https://github.com/kubernetes/kubernetes/blob/v1.24.0-alpha.1/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/helpers.go#L228

> 1. [Proposed change] In the Pod GC Controller, part of the kube-controller-manager, add a new function called gcTerminating. This function would need to go through all the Pods in terminating state, verify that the Node is NotReady. If so, do the following:

Reviewer 7: I see. Thank you.

  1.  Upon seeing the `out-of-service` taint, the Pod GC Controller will forcefully delete the pods on the node if there are no matching tolation on the pods.

>   1. new `out-of-service` taint has NoExecute effect, meaning the pod will be evicted and a new pod will not schedule on the shutdown node unless it has a matching toleration. For example, `out-of-service=nodeshutdown:NoExecute` or `out-of-service=hardwarefailure:NoExecute`.

Reviewer 11: I think you can merge this with the above mention of the taint.

Reviewer: Done

>   1. We'll follow taint and toleration policy. If a pod is set to tolerate all taints and effects, that means user does NOT want to evict pods when node is not ready. So GC controller will filter out those pods and only forcefully delete pods that do not have a matching toleration.

Reviewer 11: Earlier you said this only affects terminating pods, so I think the thing to do is emphasize that.  If your pod tolerates `out-of-service`, then it will not be terminated by the taint logic, therefore none of this applies.

Reviewer: Added.

In this KEP, we are proposing a way to handle node shutdown cases that are not detected by Kubelet's Node Shutdown Manager. The pods will be forcefully deleted in this case, trigger the deletion of the VolumeAttachments, and new pods will be created on a new running node so that application can continue to function.

Similarly, this approach can also be applied to the case when a node is in a non-recoverable state such as the hardware failure or broken OS.

### Use Cases

* User wants a stateful workload to failover to another running node in the following cases with minimum intervention when it's safe to do so.

>   * The node is shutdown

Reviewer 2: maybe here we can add some more details about use cases. 
1. If user wants to intentionally shutdown node, he/she can validate whether graceful node shutdown feature works. Kubelet is able to detect node is shutting down and gracefully delete pods. 
If graceful shutdown is not working or node is in non-recoverable state, user now can enable this feature, and add "out-of-service=nodeshutdown:NoExecute" taint which will be explained in detail below to trigger non-graceful shutdown behavior.

Reviewer: Sure

Reviewer: Done.

Existing logic:
1. When a node is not reachable from the control plane, the health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to False or Unknown (unreachable) if lease is not renewed for a specific grace period. Node Status becomes NotReady.

1. After 300 seconds (default), the Taint Manager tries to delete Pods on the Node after detecting that the Node is NotReady. The Pods will be stuck in terminating status.

Proposed logic change:
1. [Proposed change] This proposal requires a user to apply a `out-of-service` taint on a node when the user has confirmed that this node is shutdown or in a non-recoverable state due to the hardware failure or broken OS. Note that user should only add this taint if the node is not coming back at least for some time. If the node is in the middle of restarting, this taint should not be used.

1. [Proposed change] In the Pod GC Controller, part of the kube-controller-manager, add a new function called gcTerminating. This function would need to go through all the Pods in terminating state, verify that the node the pod scheduled on is NotReady. If so, do the following:

>   1.  Upon seeing the `out-of-service` taint, the Pod GC Controller will forcefully delete the pods on the node if there are no matching tolation on the pods. This new `out-of-service` taint has NoExecute effect, meaning the pod will be evicted and a new pod will not schedule on the shutdown node unless it has a matching toleration. For example, `out-of-service=nodeshutdown:NoExecute` or `out-of-service=hardwarefailure:NoExecute`.

Reviewer 2: mention to suggest use "NoExecute" effect in taint to make sure pods will be evicted (deleted) and fail over to other nodes?

Reviewer: Sure.

Reviewer: Added.

Kubernetes are observable, scalable and supportable; can be safely operated in
production environments, and can be disabled or rolled back in the event they
cause increased failures in production. See more in the PRR KEP at
https://git.k8s.io/enhancements/keps/sig-architecture/1194-prod-readiness.

The production readiness review questionnaire must be completed and approved
for the KEP to move to `implementable` status and be included in the release.

In some cases, the questions below should also have answers in `kep.yaml`. This
is to enable automation to verify the presence of the review, and to reduce review
burden and latency.

The KEP must have a approver from the
[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
team. Please reach out on the
[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
you need any help or guidance.

-->

### Feature Enablement and Rollback

_This section must be completed when targeting alpha to a release._

* **How can this feature be enabled / disabled in a live cluster?**
  - [ ] Feature gate (also fill in values in `kep.yaml`)
    - Feature gate name: NonGracefulFailover

>     - Components depending on the feature gate: kube-controller-manager

Reviewer 15: it surprises me that there isn't work to do on the kubelet side for handling when a kubelet that was previously thought lost forever reconnects.

Reviewer: This KEP does not propose any changes in Kubelet. The proposed changes are in the Pod CG Controller and Attach/detach controller.
Since the pods are forcefully deleted from the shutdown node and VolumeAttachments are also deleted before the node comes up and accept new pods to be scheduled on it again, we don't need to make changes in Kubelet.
The current KEP does rely on the user to apply the taint when the node is down and remove the taint when the node is recovered. In the future, we'll make this more automatic.

Reviewer 16: One potential problem I see with this is user applies the out-of-service taint, the pod is rescheduled and the volume attachments are removed, and later the pod is recovered. Correct me if I'm wrong, but when the node comes back and the user subsequently removes the out-of-service taint, if the node was not rebooted during that period (perhaps it only had loss of connectivity), theire might be remaining mounts or paths on the node for the volumes that were not cleaned up because NodeUnpublishVolume and NodeUnstageVolume were not called. If you subsequently force another reschedule the pods that would result in in trying to create a replace pod back on the original node, it may fail to mount the volumes, depending on how the CSI driver views these existing mounts. 
So is a reboot of the node required during the period of application of the out-of-service taint to removal of the out-of-service taint?

Reviewer 13: i agree, there is no change in the kubelet.

this is really about the standard convention for the taint, and making the volume controller respect it.

Reviewer 2: shutdown case means the node should be rebooted when it comes back. For other non-recoverable cases like hardware failure, I think OS reboot will be part of repair? We can make this point clear in the doc.

Reviewer: Added a note in the KEP.

  - [Risks and Mitigations](#risks-and-mitigations)
- [Design Details](#design-details)
  - [Test Plan](#test-plan)
  - [Unit tests](#unit-tests)
  - [E2E tests](#e2e-tests)
  - [Graduation Criteria](#graduation-criteria)
    - [Alpha](#alpha)
    - [Alpha -&gt; Beta Graduation](#alpha---beta-graduation)
    - [Beta -&gt; GA Graduation](#beta---ga-graduation)
  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
- [Production Readiness Review Questionnaire](#production-readiness-review-questionnaire)
  - [Feature Enablement and Rollback](#feature-enablement-and-rollback)
  - [Rollout, Upgrade and Rollback Planning](#rollout-upgrade-and-rollback-planning)
  - [Monitoring Requirements](#monitoring-requirements)
  - [Dependencies](#dependencies)
  - [Scalability](#scalability)
  - [Troubleshooting](#troubleshooting)
- [Implementation History](#implementation-history)
- [Alternatives](#alternatives)
  - [rely on the taint as a lock mechanism (taint and untaint by the same actor)](#rely-on-the-taint-as-a-lock-mechanism-taint-and-untaint-by-the-same-actor)
  - [Node fencing](#node-fencing)
<!-- /toc -->

> ## Release Signoff Checklist

Reviewer 5: we need also `Production Readiness Review Questionnaire` ? https://github.com/kubernetes/enhancements/tree/master/keps/NNNN-kep-template#production-readiness-review-questionnaire
according to this comment from 1.23 Enhancements team: https://github.com/kubernetes/enhancements/issues/2268#issuecomment-910732554

* Automate self-healing for stateful workloads.
* This proposal only target CSI volumes.

### Non-Goals

* Node/control plane partitioning other than a node shutdown is not covered by the proposal, but will be addressed in the future and built on top of this design.
* Implement in-cluster logic to handle node/control plane partitioning. 
* Enable detach for all the storage providers.
* Existing in-tree volumes are not targeted by this proposal.

## Proposal

This proposal introduces a new option in CSIDriver. If set to true, it would trigger volume detach if a node is shutdown.

* In the [CSIDriverSpec](https://github.com/kubernetes/kubernetes/blob/v1.20.2/pkg/apis/storage/types.go#L266), add a `SafeDetach` boolean to indicate if it is safe to detach.

```go
// CSIDriverSpec is the specification of a CSIDriver.
type CSIDriverSpec struct {
  ...

  // SafeDetach indicates this CSI volume driver is able to perform detach
  // operations. When set to true, the CSI volume driver needs to ensure that 
  // ControllerUnpublishVolume() implementation is checking whether a volume
  // can be detached safely
  // +optional
  SafeDetach *bool
}
```

> * The Graceful Node Shutdown [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2000-graceful-node-shutdown) introduced a way to detect a node shutdown. When the NodeShutdownManager detects a node is shutting down, it sets the `Reason` field of a `PodStatus` to a constant named `nodeShutdownReason`. Currently `nodeShutdownReason` is in the `nodeshutdown` package. We need to export this constant by renaming it to `NodeShutdownReason` and move it to a package that can be imported by the Pod GC Controller.

Reviewer 17: I think there is some confusion to Graceful Node Shutdown. Graceful Node Shutdown largely thinks the node is going down for a reboot and will come back; not be shutdown permanently. If a node is upgrading, then this proposal would detach all the volumes, then cause the Kubelet to thundering-heard attach the volumes once the node comes back up.

Another issue will be network partitions. The Kubelet and Control Plane are designed to be resilient to network partitions. This proposal side steps the network partition question, which is a valid consideration since all pods are assumed to still be running on a Partitioned node. This Partitioned Node could still be writing to a volume.

Regarding
> After 300 seconds (default), the Taint Manager tries to delete Pods on the Node after detecting that the Node is NotReady. The Pods will be stuck in terminating status.

The pods in the API may be stuck in terminating state, but the pods could still be running. Once the network partition is resolved, then the API statuses will be updated correctly. The scheduler will then make a determination if the pods on the resolved node should be terminated.

Reviewer 12: > Graceful Node Shutdown largely thinks the node is going down for a reboot and will come back; not be shutdown permanently.

I'm not sure if this is correct assumption. I think with graceful node shutdown we don't really know if the node is going to reboot or will shutdown permanently. All graceful node shutdown knows is that there is a system shutdown imminent, and it delays it to terminate pods gracefully. 

/cc @mrunalp

Reviewer 17: True. After thinking about it overnight, that point of mine does not make sense, but the others are valid.

  1. [Proposed change] After applying taint on the shutdown node, forcefully delete pods that are stuck.

1. [Proposed change] Once pods are selected and forcefully deleted, the attachdetach reconciler should check if the `SafeDetach` flag in CSIDriver is set to true. If so, it should skip the `attachedVolume.MountedByNode` check for the volumes backing these pods and allow `volumeAttachement` to be deleted.

1. This would trigger the deletion of the `volumeAttachement` objects. This would allow `ControllerUnpublishVolume` to happen before `NodeUnpublishVolume` and/or `NodeUnstageVolume` are called.

1. When the `external-attacher` detects the `volumeAttachement` object is being deleted, it calls CSI driver's `ControllerUnpublishVolume`.

1. [Proposed change] The CSI driver would need to ensure that no volume is being used through a mount point. If unable to determine the usage, the CSI driver can return an error without detaching the volume. The `external-attacher` would then retry with backoffs.

### Handle the Return of Shutdown Node

> If the nodes that were previously shutdown with pods forcefully deleted come back, we need to make sure they are cleaned up before pods are scheduled on them again.

Reviewer 2: the scope of the KEP now is changed. It requires user to make sure node is down already which means no pods should run at this point.

> If the nodes that were previously shutdown with pods forcefully deleted come back, we need to make sure they are cleaned up before pods are scheduled on them again.

Reviewer 11: So if we can't CONFIRM it is powered off, we still let it hang, right?

> If the nodes that were previously shutdown with pods forcefully deleted come back, we need to make sure they are cleaned up before pods are scheduled on them again.

Reviewer: If the quarantine taint was not added to the shutdown node, the pods will be stuck in terminating status forever if the shutdown node remains down. If the shutdown node comes up, however, the original volumeAttachment will be deleted and the pods will be moved to a new node successfully. Added this to the KEP to clarify.

> If the nodes that were previously shutdown with pods forcefully deleted come back, we need to make sure they are cleaned up before pods are scheduled on them again.

Reviewer 18: I have tried if we use following configs in statefulset, the pods would be in stuck in terminating status for ~7min, what's the benefit if user already sets  following configs? Manually add a taint to a node in problem is still not an ideal way IMO:
```yaml
    spec:
      tolerations:
      - key: "node.kubernetes.io/not-ready"
        operator: "Exists"
        effect: "NoExecute"
        tolerationSeconds: 10
      - key: "node.kubernetes.io/unreachable"
        operator: "Exists"
        effect: "NoExecute"
        tolerationSeconds: 10
      terminationGracePeriodSeconds: 0
```
check details here: 
https://github.com/kubernetes-sigs/azuredisk-csi-driver/tree/master/docs/known-issues/node-shutdown-recovery

> If the nodes that were previously shutdown with pods forcefully deleted come back, we need to make sure they are cleaned up before pods are scheduled on them again.

Reviewer: @andyzhangx What K8s version have you tested this with?  Also which command did you use to shutdown the node?  Just want to check if this goes thru the Graceful Node Shutdown code path.

> If the nodes that were previously shutdown with pods forcefully deleted come back, we need to make sure they are cleaned up before pods are scheduled on them again.

Reviewer 18: @xing-yang tested with ~1.19, and I powered off the node directly.

> If the nodes that were previously shutdown with pods forcefully deleted come back, we need to make sure they are cleaned up before pods are scheduled on them again.

Reviewer 2: @andyzhangx I checked your link. In your workaround, why those tolerations are needed? When node is shutdown, pod is stuck in terminating and that's why disk will not be detached from the node. 
When setting terminationGracePeriodSeconds to 0,  will it make a difference so that pod will be  force deleted from API server even if the node is down (kubelet will not response)?

> If the nodes that were previously shutdown with pods forcefully deleted come back, we need to make sure they are cleaned up before pods are scheduled on them again.

Reviewer 18: @jingxu97 I tried again with `k apply -f https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/master/docs/known-issues/node-shutdown-recovery/statefulset-azuredisk.yaml` on k8s v1.20.9 cluster, after I power off node `aks-agentpool-35915375-vmss000000`, you could see `statefulset-azuredisk-0` is stuck in `Terminating` forever, while `statefulset-autorecover-0` would be terminated and scheduled to node `aks-agentpool-35915375-vmss000001` after a few minutes. I think `terminationGracePeriodSeconds: 0` makes the difference, with that setting, pod would be terminated immediately in pod eviction when node `aks-agentpool-35915375-vmss000000` becomes `NotReady`, I think you could get same test result.

So here the point is `terminationGracePeriodSeconds: 0` already have same effect, do we really need this KEP since this KEP requires adding a taint manually on the node in problem.

```console
# k get po -o wide
NAME                                    READY   STATUS              RESTARTS   AGE     IP             NODE                                NOMINATED NODE   READINESS GATES
statefulset-autorecover-0               1/1     Running             0          10m     10.240.0.130   aks-agentpool-35915375-vmss000001   <none>           <none>
statefulset-azuredisk-0                 1/1     Terminating         0          28m     10.240.0.35    aks-agentpool-35915375-vmss000000   <none>           <none>
```

> If the nodes that were previously shutdown with pods forcefully deleted come back, we need to make sure they are cleaned up before pods are scheduled on them again.

Reviewer 2: @andyzhang To set terminationGracePeriodSeconds, I think user need to add this to every pod spec. For system admin, they will not be able to do that.

Also in normal cases, we might not recommend to set terminationGracePeriodSeconds to 0 since this will not allow pod to gracefully terminate even on healthy node which might cause undesired behavior. 

The goal of this KEP is to recover form node shutdown automatically. As the first phase, we still require minimum human intervention, in this case, to add a taint to the node. In the future, this could be done by controller automatically.

  - [ ] Feature gate (also fill in values in `kep.yaml`)
    - Feature gate name: NonGracefulFailover
    - Components depending on the feature gate: kube-controller-manager, kubelet
  - [ ] Other
    - Describe the mechanism:
    - Will enabling / disabling the feature require downtime of the control
      plane?
      Yes.
    - Will enabling / disabling the feature require downtime or reprovisioning
      of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).
      No.

* **Does enabling the feature change any default behavior?**
  <!--
  Any change of default behavior may be surprising to users or break existing
  automations, so be extremely careful here.
  -->
  Yes. If this feature is enabled, the pod gc controller in kube-controller-manager
  would need to ensure that `SafeDetach` option in `CSIDriver` is true before
  forcefully deleting pods.

* **Can the feature be disabled once it has been enabled (i.e. can we roll back
  the enablement)?**
  Also set `disable-supported` to `true` or `false` in `kep.yaml`.
  Describe the consequences on existing workloads (e.g., if this is a runtime
  feature, can it break the existing applications?).

>   Yes. If this feature is disable once it has been enabled, it falls back to

Reviewer 15: What happens if this feature is disabled part way the process?  For instance, if the node has been tainted, will the automatic untainting happen or will it be stuck?

It's ok for some aspects to require manual interaction, but we need to enumerate what is required.

Reviewer 7: The current design says no automatic taint-removing feature.

  Try to be as paranoid as possible - e.g., what if some components will restart
   mid-rollout?

* **What specific metrics should inform a rollback?**

* **Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?**
  Describe manual testing that was done and the outcomes.
  Longer term, we may want to require automated upgrade/rollback tests, but we
  are missing a bunch of machinery and tooling and can't do that now.

* **Is the rollout accompanied by any deprecations and/or removals of features, APIs,
fields of API types, flags, etc.?**
  Even if applying deprecation policies, they may still surprise some users.

### Monitoring Requirements

_This section must be completed when targeting beta graduation to a release._

* **How can an operator determine if the feature is in use by workloads?**
  Ideally, this should be a metric. Operations against the Kubernetes API (e.g.,
  checking if there are objects with field X set) may be a last resort. Avoid
  logs or events for this purpose.

* **What are the SLIs (Service Level Indicators) an operator can use to determine
the health of the service?**

>   - [ ] Metrics

Reviewer 15: A reminder that you'll need to have metrics for beta.

Reviewer: Sure.

- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
- [ ] Graduation criteria is in place
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those
approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).

**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
[kubernetes/website]: https://github.com/kubernetes/website

## Summary

> In case of a node being shutdown (i.e. not running) the control plane doesn't make the right

Reviewer 4: This kep has changed focus many times so it's difficult to remember what use cases we're trying to target now. Can you add a "Use cases" section that outlines what kind of node shutdown scenarios we're trying to address?

assumptions to enable stateful workload to fail-over. This KEP introduces a flow to do so.

> ## Motivation

Reviewer 4: > The Graceful Node Shutdown KEP talks about adding ShutdownGracePeriod for Pods. It didn't talk about how to handle VolumeAttachment.

I would like to double click on this. Reading the graceful node shutdown blog, it looks like what the feature does is delay the shutdown of a node to give a longer grace period for pods to gracefully terminate. If that's the case, I would expect that grace period to also allow time to unmount and detach volumes from the node.

If that's not working, then I agree with @mkimuram that this may be a bug in the feature, and we should investigate how to fix it. 

This proposal seems to be handling the scenario where perhaps the grace period has passed and the node has ended up shutting down before we are able to finish tearing down the volumes. But I think in terms of priority, we should focus on getting the graceful sequence to work first.

Reviewer: It is not that we don't have enough time to detach the volumes. The detach won't happen unless if we make the code changes suggested in this KEP. I've tested this by shutting down one node and observing the pods of a StatefulSet won't move to another node, even after enabling the Graceful Node Shutdown feature (and also set those two periods to > 0).

@bobbypage Can you chime in?

Reviewer 12: The current behavior of graceful node shutdown feature is upon shutdown kubelet terminates the pod and puts it into the failed phase. There's wasn't any special logic applied for volume detachment, but it's great to see this KEP consider how we can leverage graceful node shutdown for volume detaching upon shutdown. 

I chatted with @msau42 offline, it seemed like there's two steps in the volume detaching process:

1) kubelet unmounting the volume
2) signaling to kube-controller-manager that the volume is safe to detach

For #1, it seems like volume detachment is managed by kubelet and it will detach if the pod is deleted and containers have finished cleaning up [1]. Since the graceful node shutdown terminates the pod, I would expect eventually that kubelet volume manager will see the pod is terminated and that the detach would happen. We should confirm this is in fact the case.

For #2, as per my chat with @msau42 it depends on the pod having a deletion timestamp set which graceful node shutdown does not currently do. 

[1] https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go#L244

Reviewer 4: @bobbypage can you also clarify when you say "upon shutdown" do you mean "when the node gets a shutdown signal but in the grace period before the system actually goes down"

Reviewer 12: > @bobbypage can you also clarify when you say "upon shutdown" do you mean "when the node gets a shutdown signal but in the grace period before the system actually goes down"

Yes. 

The logic is that kubelet is watching system shutdown events from systemd. When kubelet is notified of the shutdown event, it uses a systemd inhibitor lock to request additional time before the system gets shutdown. This duration of additional grace period is configured in KubeletConfig as `ShutdownGracePeriod`. Kubelet uses this additional grace period to gracefully terminate pods.

Reviewer 8: > Since the graceful node shutdown terminates the pod, I would expect eventually that kubelet volume manager will see the pod is terminated and that the detach would happen. We should confirm this is in fact the case.

Current behavior seems just a bug in k8s 1.22 and it worked well with k8s 1.21. 
Please see my comments [here](https://github.com/kubernetes/kubernetes/issues/104531#issuecomment-912717050) and [here](https://github.com/kubernetes/kubernetes/issues/104531#issuecomment-912822663).

Reviewer 11: > For #2, as per my chat with @msau42 it depends on the pod having a deletion timestamp set which graceful node shutdown does not currently do. During graceful node shutdown, kubelet terminates the pod and puts the pod into failed phase, but does not delete the pod or set deletion timestamp.

How "eventually" ?  I've heard from many people that this long delay is a problem.

Reviewer 4: For the graceful node shutdown case, it is expected that volume teardown happens with the same time expectations as a normal pod deletion. It does not appear to be happening in this case, so we're focusing right now on understanding the graceful node shutdown design in more detail and its impact on kubelet volume manager and AD controller.

Reviewer: There are two main issues with the Graceful Node Shutdown feature:
1. Whether Kubelet can detect a “Graceful node shutdown” depends on which shutdown command is used. In my testing on Ubuntu, “systemctl poweroff” is the only command that is detected by Kubelet as a “Graceful node shutdown”. Other commands such as “shutdown -h now” or shutting down a VM from vSphere UI are not detected by Kubelet, so they all fall into “Non-graceful node shutdown” cases.
2. The bug in k8s 1.22 mentioned above. In k8s 1.21, Pod status is changed to “Failed” when Kubelet detects the node shutdown so volumes are detached properly and pods are moved to a remaining running node. In k8s 1.22, the Pod status is not changed to “Failed” and stays as “Running” when Kubelet detects the node shutdown, so volumes stay as attached and pods are not moved to a remaining running node.

* In the [CSIDriverSpec](https://github.com/kubernetes/kubernetes/blob/v1.20.2/pkg/apis/storage/types.go#L266), add a `SafeDetach` boolean to indicate if it is safe to detach.

```go
// CSIDriverSpec is the specification of a CSIDriver.
type CSIDriverSpec struct {
  ...

  // SafeDetach indicates this CSI volume driver is able to perform detach
  // operations. When set to true, the CSI volume driver needs to ensure that 
  // ControllerUnpublishVolume() implementation is checking whether a volume
  // can be detached safely
  // +optional
  SafeDetach *bool
}
```

* The Graceful Node Shutdown [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2000-graceful-node-shutdown) introduced a way to detect a node shutdown. When the NodeShutdownManager detects a node is shutting down, it sets the `Reason` field of a `PodStatus` to a constant named `nodeShutdownReason`. Currently `nodeShutdownReason` is in the `nodeshutdown` package. We need to export this constant by renaming it to `NodeShutdownReason` and move it to a package that can be imported by the Pod GC Controller.

Existing logic:

> * When a node is partitioned from the control plane, the health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to False or Unknown (unreachable) if lease is not renewed for a specific grace period. Node Status becomes NotReady.

Reviewer 4: nit: since this is an ordered sequence, can you use a numbered list?

* After 300 seconds (default), the Taint Manager tries to delete Pods on the Node after detecting that the Node is NotReady. The Pods will be stuck in terminating status.

Proposed logic change:

> * The Pod GC Controller, part of the kube-controller-manager, would need to go through all the Pods in terminating state, verify that the Node is NotReady, check if the `Reason` in the `PodStatus` is set to the `NodeShutDownReason` constant, and then check if the `SafeDetach` flag in CSIDriver is set to true. If so, forcefully delete pods that are stuck.

Reviewer 4: For each of these steps, can you clarify what is existing logic and what is new?

Reviewer: Added a [Proposed change] in front of each changed step.

> * Specifically the Pod GC Controller (the only controller able to forcefully delete pods) would also check now for stateful pods, and select those that satisfy ALL of the following conditions:

Reviewer 4: selects those pods and do what?

Reviewer: This is actually part of the above step.  I'll remove this.

  * pods that are backed by a CSI driver that supports `SafeDetach`.
  * pods that are backed by volumes that aren't marked as `ReadWriteMany`.

> * Once pods are selected and forcefully deleted, the attachdetach reconciler should check if the `SafeDetach` flag in CSIDriver is set to true. If so, it should skip the `attachedVolume.MountedByNode` check for the volumes backing these pods and allow `volumeAttachement` to be deleted.

Reviewer 4: Force deleting a pod changes the expected pod lifecycle only for Pods that use a CSI volume. It seems very special cased. I think we should discuss with SIG-Node on how to make this more generic behavior.

Reviewer: @bobbypage Please take a look of this step and provide comments.

* This would trigger the deletion of the `volumeAttachement` objects. This would allow `ControllerUnpublishVolume` to happen before `NodeUnpublishVolume` and/or `NodeUnstageVolume` are called.

* When the `external-attacher` detects the `volumeAttachement` object is being deleted, it calls CSI driver's `ControllerUnpublishVolume`. The CSI driver would need to ensure that no volume is being used through a mount point. If unable to determine the usage, the CSI driver can return an error without detaching the volume. The `external-attacher` would then retry with backoffs.

> * The Pod GC Controller would also apply a `quarantine` taint on the nodes whose pods are forcefully deleted due to the shutdown. This should happen before the pods are being evicted. The "quarantine" taint indicates that new pods should not be scheduled on the node.

Reviewer 4: Can you move this earlier in the proper sequence order?

Reviewer: Moved up.

### Handle the Return of Shutdown Node

If the nodes that were previously shutdown with pods forcefully deleted come back, we need to make sure they are cleaned up before pods are scheduled on them again.

The health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to True if lease is renewed again. Node Status becomes Ready.

> When the node comes up, Kubelet volume manager reconciler will try to clean up mounts by calling `NodeUnpublishVolume`. Kubelet volume manager reconciler will also try to reconstruct global mount for CSI volumes and call `NodeUnstageVolume`. If there isn't enough information to reconstruct global mount, Kubelet volume manager should cleanup the [global mount directory](/var/lib/kubelet/plugins/kubernetes.io/csi/pv/{pvname}/globalmount) to avoid problems when new pods are scheduled to this node again.

Reviewer 4: When volume reconstruction runs, is it guaranteed that kubelet pod manager is updated with the latest API state? I remember kubelet has implemented checkpointing to be able to recover state after a restart.

Reviewer 4: If reconstruction is not able to cleanup properly, what should we do?

Reviewer: It looks like sig-node didn't want to go with Kubelet checkpoint.  See comments here:
https://github.com/kubernetes/kubernetes/issues/489#issuecomment-868768740
The Kubelet checkpoint-restore KEP has been updated to address a different issue - pod migration between nodes:
https://github.com/kubernetes/enhancements/pull/1990

> The Pod GC Controller should also remove the `quarantine` taint from the nodes that are already cleaned up and are safe to run pods again. We need to make sure the taint is only removed after detach is finished.

Reviewer 4: How does Pod GC controller know when a node is done cleaning up?

Reviewer 4: "only removed after detach is finished": does this imply that Pod GC controller needs to be watching VolumeAttachment objects?

Reviewer: Good question.  We probably should not have Pod GC Controller watch VolumeAttachment if not necessary. That means we probably should have Kubelet volume manager removes the taint after reconstruction.  I'll update the KEP.  Let me know if you have other thoughts.
